fedimint-server/src/lib.rs

```diff
@@ -352,6 +367,10 @@ impl FedimintServer {
             }
         }
 
+        if self.consensus.is_at_shutdown_threshold().await {
+            self.task_group.shutdown().await;
```

Comment:
Did you test if this works? Some tasks were not adhering to the shutdown signal yet …

Reply:
BTW. My consensus processing logic here is limited so I'm not sure if there's anything to do anymore about it, but:

We should gate the epoch processing on lack of blocking shutdown item explicitly and not relay on task shutdown system to enforce it because it might be replaced / refactored / buggy.

The consensus shutdown should be somewhere in the most common code before processing the epoch height scheduled for shutdown (in `self.consensus.process_consensus_outcome`?). *Before*, so it can be never accidentally skipped. If it's after then one can process the epoch, write state to the db, get restarted and then never run the check altogether, potentially processing more epochs than other peers, etc. Or some corner case (rejoining consensus, etc.) can run through bunch of epochs without doing the check etc.

Reply:
If it's before then we will ignore any other items in that epoch?  That might cause issues for clients who think their transactions were accepted.

Reply:
We could process the items in the shutdown epoch and the shutdown id will be persisted in the same DB transaction.

As a fail-safe we can check for the non-existence of the shutdown id upon start-up or epoch processing (would need to be manually removed by passing it in as a start-up parameter)

Reply:
> If it's before then we will ignore any other items in that epoch? That might cause issues for clients who think their transactions were accepted.

We will exit.

So an epoch that had consensus item that made the shutdown reach a threshold is the last one that was processed, and it will update the db (atomically: in the same db transactions as all other side-effects of that epoch). And then the next epoch won't be able to process due to the check right before it.